subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,792 @@
+  }
+
+  return TRUE;
+}
+
+static GstFlowReturn
+gst_rtp_h266_pay_handle_buffer (GstRTPBasePayload * rtpbasepay,
+    GstBuffer * buffer)
+{
+  GstRtpH266Pay *rtph266pay = GST_RTP_H266_PAY (rtpbasepay);
+  GstAdapter *adapter = rtph266pay->adapter;
+  gsize adapter_size;
+
+  GST_DEBUG_OBJECT (rtph266pay, "New buffer %" GST_PTR_FORMAT, buffer);
+  gst_adapter_push (adapter, buffer);
+
+  // Advance until the first start code to skip heading zeroes
+  adapter_size = gst_adapter_available (adapter);
+  gssize next_nalu_start =
+      gst_adapter_masked_scan_uint32 (adapter, NALU_SC_MSK, NALU_SC_VAL, 0,
+      adapter_size);
+  if (next_nalu_start < 0) {
+    GST_WARNING_OBJECT (rtph266pay, "Not NALU found");
+    gst_adapter_flush (adapter, adapter_size);
+    return GST_FLOW_OK;
+  }
+  gst_adapter_flush (adapter, next_nalu_start);
+
+  // For each NALU
+  while ((adapter_size = gst_adapter_available (adapter)) > NALU_SC_LEN) {
+    // Find next start code, skipping the actual one
+    gssize next_nalu_start =
+        gst_adapter_masked_scan_uint32 (adapter, NALU_SC_MSK, NALU_SC_VAL,
+        NALU_SC_LEN, adapter_size - NALU_SC_LEN);
+
+    // If no start code found, the length of the NALU is the remaining size
+    gssize nalu_len = (next_nalu_start < 0) ? adapter_size : next_nalu_start;
+
+    if (nalu_len <= NALU_SC_LEN) {
+      GST_WARNING_OBJECT (rtph266pay, "NALU too small %ld, skipping it",
+          nalu_len);
+      gst_adapter_flush (adapter, nalu_len);
+      continue;
+    }
+
+    // Remove the start code
+    nalu_len -= NALU_SC_LEN;    // Here, it'll always be > 0
+    gst_adapter_flush (adapter, NALU_SC_LEN);
+
+    GstBuffer *nalu_buf = gst_adapter_take_buffer (adapter, nalu_len);
+    GST_BUFFER_PTS (nalu_buf) = gst_adapter_prev_pts (adapter, NULL);
+    GST_BUFFER_DTS (nalu_buf) = gst_adapter_prev_dts (adapter, NULL);
+    gst_rtp_h266_pay_process_nalu (rtph266pay, nalu_buf);
+  }
+
+  return gst_rtp_h266_pay_push_pending_data (rtph266pay, FALSE);
+}
+
+static gboolean
+gst_rtp_h266_pay_sink_event (GstRTPBasePayload * rtpbasepay, GstEvent * event)
+{
+  GstRtpH266Pay *rtph266pay = GST_RTP_H266_PAY (rtpbasepay);
+  GstFlowReturn ret = GST_FLOW_OK;
+
+  switch (GST_EVENT_TYPE (event)) {
+    case GST_EVENT_FLUSH_STOP:
+      gst_adapter_clear (rtph266pay->adapter);
+      g_queue_clear_full (&rtph266pay->nalus, (GDestroyNotify) nalu_free);
+      break;
+    case GST_EVENT_EOS:
+      GST_DEBUG_OBJECT (rtph266pay, "EOS: Draining");
+      ret = gst_rtp_h266_pay_push_pending_data (rtph266pay, TRUE);
+      break;
+    case GST_EVENT_STREAM_START:
+      gst_rtp_h266_pay_clear_xps_cache (rtph266pay);
+      break;
+    default:
+      break;
+  }
+
+  if (ret != GST_FLOW_OK)
+    return FALSE;
+
+  return GST_RTP_BASE_PAYLOAD_CLASS (parent_class)->sink_event (rtpbasepay,
+      event);
+}
+
+static void
+gst_rtp_h266_pay_process_nalu (GstRtpH266Pay * rtph266pay, GstBuffer * nalu_buf)
+{
+  Nalu *nalu;
+
+  g_assert (nalu_buf);
```

Comment:
I would use here `g_return_if_fail (nalu_buf != NULL);`

Reply:
:thinking: having a `NULL` buffer here would mean having a really bad bug in the code... I'll think about it.

Reply:
`GST_IS_BUFFER ()` to fail would be asserting that `gst_adapter_take_buffer()` actually returns a `GstBuffer`. IMO, this makes no sense. We must trust GStreamer is working as expected.

Reply:
I'm leaving this as is. I don't want the code to continue in case `nalu_buf == NULL`. It would mean a big bug in the parser, which could lead to security vulnerabilities.